File: src/edgemlsdk/src/docs/samples/C++/Gst/variables_pipeline.cpp
#include <nlohmann/json.hpp>

#include <Panorama/trace.h>
#include <Panorama/app.h>
#include <Panorama/gst.h>

using namespace Panorama;

void CreatePropertiesFile(const std::string& pattern)
{
    // Create a file to hold some variables.
    FILE* fptr = fopen("variables_sample.json", "w");

    nlohmann::json variables;
    
    nlohmann::json string_variable;
    string_variable["type"] = "string";
    string_variable["immutable"] = true;
    string_variable["value"] = pattern;

    variables["pattern"] = string_variable;

    fprintf(fptr, "%s", variables.dump().c_str());
    fclose(fptr);
}

int main()
{
    ADD_CONSOLE_TRACE;
    HRESULT hr = S_OK;

    // Create a file that can be loaded by the FilePropertyDelegate that contain a variable
    CreatePropertiesFile("snow");

    // Create the file property delegate to read from that file
    ComPtr<IPropertyDelegate> file_delegate;
    CHECKHR(CreateFilePropertyDelegate(file_delegate.AddressOf(), "./variables_sample.json"));

    // Create the application object.
    ComPtr<IApp> app = App::Create();
    CHECKNULL(app, E_FAIL);

    // Add the property delegate to the application object
    CHECKHR(app->AddPropertyDelegate(file_delegate));

    // Initialize the GStreamer framework
    CHECKHR(GStreamer::Initialize());

    // Create the pipeline object with a simple definition
    ComPtr<IPipeline> pipeline;
    CHECKHR(Pipeline::Create(pipeline.AddressOf(), "my-pipeline", "videotestsrc name=src pattern=${pattern} ! ximagesink", app));

    // Start the pipeline
    CHECKHR(pipeline->Start());

    // Wait for a few seconds then update the file with a new pattern
    ThreadSleep(3000);
    CreatePropertiesFile("ball");

    // Synchronize the application to update changes to properties
    ComPtr<IPropertyCollection> changed_properties; // not used in this example
    CHECKHR(app->Synchronize(changed_properties.AddressOf()));

    // Refresh the pipeline to apply property changes.  Since the value of pattern is marked as immutable it will cause the pipeline to restart.
    CHECKHR(pipeline->Refresh());

    // Wait for a few seconds then shutdown
    ThreadSleep(3000);
    CHECKHR(pipeline->Stop()); // Happens automatically when pipeline goes out of scope, here for completeness

    CHECKHR(GStreamer::Shutdown());
    return hr;
}

File: src/edgemlsdk/src/docs/samples/Python/Gst/variables_pipeline.py
#
#  Copyright 2025 Amazon Web Services, Inc.
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#

import json
import time
from panorama import trace
from panorama import gst
from panorama import application
from panorama import properties

def create_properties_file(pattern):
    data = {
        "pattern": 
        {
            "type": "string",
            "immutable": True,
            "value": pattern
        }
    }

    with open('./variables_sample.json', 'w') as file:
        json.dump(data, file, indent=4)
    
def main():
    create_properties_file("snow")
    trace.add_console_trace_listener()

    # Create the file property delegate
    file_delegate = properties.create_file_property_delegate("./variables_sample.json")

    # Create the app object
    app = application.create()
    app.add_property_delegate(file_delegate)

    # Initialize the Gstreamer framework
    gst.initialize()

    # Create the pipeline
    pipeline = gst.pipeline.create("my-pipeline", "videotestsrc name=src pattern=${pattern} ! ximagesink", app)

    # Start the pipeline
    pipeline.start()

    # Wait a few seconds and update the properties file. Since pattern is marked as immutable changing its value will cause the pipeline to restart
    time.sleep(3)
    create_properties_file("ball")

    # Synchronize the application to get changes to the properties
    app.synchronize()

    # Refresh the pipeline to apply any property updates
    pipeline.refresh()
    
    time.sleep(3)
    pipeline.stop()
    gst.shutdown()

if __name__ == "__main__":
    main()